File: rvis.py
#! /usr/bin/env python3

# single cycle datapath RISC-V 32I
# Stages of RISCV datapath:
    # fetch
    # decode
    # execute
    # memory access
    # register writeback

# import argparse
import glob
from enum import Enum

from bitstring import Bits, BitArray
from elftools.elf.elffile import ELFFile


class Opcode(Enum):
    LUI = BitArray(bin = '0b0110111') # load upper immediate
    AUIPC = BitArray(bin = '0b0010111') # Add upper immediate to PC 
    JAL = BitArray(bin= '0b1101111') # Jump and link 
    JALR = BitArray(bin = '0b1100111') # Jump and link register
    BRANCH = BitArray(bin = '0b1100011')
    LOAD = BitArray(bin = '0b0000011')
    STORE = BitArray(bin = '0b0100011')
    IMM =  BitArray(bin = '0b0010011') # Immediate instructions
    OP = BitArray(bin = '0b0110011') # Arithmetic/Logic Ops 
    MISCMEM = BitArray(bin = '0b0001111')
    SYSTEM = BitArray(bin = '0b1110011') 


class Funct3(Enum):
    JALR = BEQ = LB = \
    SB = ADDI = ADD = SUB = \
    FENCE = ECALL = EBREAK = BitArray(bin = '0b000')
    
    BNE = LH = SH = \
    SLLI = SLL = FENCEI = CSRRW = BitArray(bin = '0b001')

    BLT = LBU = XORI = XOR = BitArray(bin = '0b100')

    BGE = LHU = SRLI = SRAI = \
    SRL = SRA = CSRRWI = BitArray(bin = '0b101')

    BLTU = ORI = OR = CSRRSI =  BitArray(bin = '0b110')

    BGEU = ANDI = AND = CSRRCI =  BitArray(bin = '0b111')
    
    LW = SW = SLTI = \
    SLT = CSRRS = BitArray(bin = '0b010')
    
    SLTIU = SLTU = CSRRC =  BitArray(bin = '0b011')


class Funct7(Enum):
    SRAI = SUB = SRA = BitArray(bin = '0b0100000')

    SLLI = SRLI = ADD = \
    SLL = SLT = SLTU = \
    XOR = SRL = OR = AND = BitArray(bin = '0b0000000')


class Regfile():
    def __init__(self):
        self.registers = [0] * 33 # x0-x31 and PC

    def __getitem__(self, key):
        return self.registers[key]
    
    def __setitem__(self, key, value):
        if key == 0: # x0 register is always 0
            return
        else:
            self.registers[key] = value & 0xFFFFFFFF # store only the lowest 4 bytes


# GLOBAL VARIABLES
regfile = None
memory = None 
PC = 32 # PC is at index 32 in the regfile

def reset():
    global regfile, memory
    regfile = Regfile()
    # 16 MB memory, byte addressable, starting at 0x8000000
    memory = bytearray(0x1000000)


def loader(data, address):
    address -= 0x80000000 # subtract offset to get back to 0 index
    if address < 0 or address > len(memory):
        raise Exception("read out of bounds")
    memory[address:len(data)] = data


def hart_dump():
    reg_names = \
        ['x0', 'ra', 'sp', 'gp', 'tp'] + ['t%d'%i for i in range(0,3)] + ['s0', 's1'] +\
        ['a%d'%i for i in range(0,8)] +\
        ['s%d'%i for i in range(2,12)] +\
        ['t%d'%i for i in range(3,7)] + ["PC"]
    file = [] 
    for i in range(len(reg_names)):
        if i != 0 and i % 8 == 0:
            file += "\n"
        if type(regfile[i] == int) and i == 32:
            file += " %4s:0x%08x" % (reg_names[i], regfile[i])
        else:
            file += " %4s:0x%08x" % (reg_names[i], regfile[i])
    print(''.join(file), '\n')


def sign_extend(sign, bits): 
    return Bits(bool = sign)*(32 - len(bits)) + bits


def fetch(address):
    address -= 0x80000000 
    return BitArray(uint = int.from_bytes(memory[address:address+4], byteorder = "little"), length = 32)


def decode(instruction):
    opcode = Opcode(instruction[25:]) # 7 bits
    rd = instruction[20:25] # set up write
    src1 = instruction[12:17] # read rs1 into temp regsiter 
    src2 = instruction[7:12] # read rs2 into temp register
    funct3 = instruction[17:20]
    funct7 = instruction[0:7]
    imm_I = sign_extend(instruction[0], instruction[0:12])
    imm_S = sign_extend(instruction[0], instruction[0:7] + instruction[20:25])
    imm_B = sign_extend(instruction[0], (bin(instruction[0]) + bin(instruction[24]) + instruction[1:7] + instruction[20:24]) << 1) # shift left to access even locations only
    imm_U = instruction[0:20] + BitArray('0x000') 
    imm_J = sign_extend(instruction[0], (bin(instruction[0]) + instruction[12:20] + bin(instruction[11]) + instruction[1:11]) << 1) 
    #print(instruction.bin, opcode)
    return opcode, rd, src1, src2, funct3, funct7, imm_I, imm_S, imm_B, imm_U, imm_J
  

def execute(opcode, rs1, rs2, funct3, funct7, imm_I, imm_S, imm_B, imm_U, imm_J, returnPC):
    
    def arithmetic(funct3, src1, src2, funct7=None):
        if funct3 == Funct3.ADD: # Add, AddI, Sub
            return src1 - src2 if funct7 == Funct7.SUB else src1 + src2
        if funct3 == Funct3.AND: # And, AndI
            return src1 & src2
        if funct3 == Funct3.OR: # Or, OrI
            return src1 | src2
        if funct3 == Funct3.XOR: # Xor, XorI
            return src1 ^ src2
        if funct3 == Funct3.SRL: # srl, srli, sra, srai
            return src1 >> (src2 % 32) if funct7 == Funct7.SRL else BitArray(hex(src1)).int >> (src2 % 32) # sra, srai 
        if funct3 == Funct3.SLL: # sll, slli
            return src1 << (src2 % 32) # max shift amount is in range [0,31] bits
        if funct3 == Funct3.SLT: # slt, slti
            #print("src1: ", src1, "src2: ", src2)
            src1 = BitArray(hex(src1 & 0xFFFFFFFF))
            src2 = BitArray(hex(src2 & 0xFFFFFFFF))
            src1 = BitArray(sign_extend(0, src1))
            src2 = BitArray(sign_extend(0, src2))
            return 1 if src1.int < src2.int else 0 
        if funct3 == Funct3.SLTU:
            #print("src1: ", src1, "src2: ", src2)
            src1 = BitArray(hex(src1 & 0xFFFFFFFF))
            src2 = BitArray(hex(src2 & 0xFFFFFFFF))
            src1 = BitArray(sign_extend(0, src1))
            src2 = BitArray(sign_extend(0, src2))
            return 1 if src1.uint < src2.uint else 0 

    def logic(funct3, src1, src2, imm_B, returnPC):
        if funct3 == Funct3.BEQ:
            return returnPC + imm_B if src1.int == src2.int else returnPC + 4 
        if funct3 == Funct3.BNE:
            return returnPC + imm_B if src1.int != src2.int else returnPC + 4 
        if funct3 == Funct3.BLT:
            return returnPC + imm_B if src1.int < src2.int else returnPC + 4 
        if funct3 == Funct3.BGE:
            return returnPC + imm_B if src1.int >= src2.int else returnPC + 4 
        if funct3 == Funct3.BLTU:
            return returnPC + imm_B if src1.uint < src2.uint else returnPC + 4
        if funct3 == Funct3.BGEU:
            return returnPC + imm_B if src1.uint >= src2.uint else returnPC + 4
            
    if opcode == Opcode.LUI:
        ALUOut = (imm_U).int
    elif opcode == Opcode.AUIPC:
        ALUOut = returnPC + imm_U.int 
    elif opcode == Opcode.JAL:
        ALUOut = returnPC + imm_J.int
    elif opcode == Opcode.JALR:
        ALUOut = int(arithmetic(Funct3.ADD, regfile[rs1.uint], imm_I.int))
    elif opcode == Opcode.BRANCH:
        funct3 = Funct3(funct3)
        ALUOut = int(logic(funct3, BitArray(hex(regfile[rs1.uint])), BitArray(hex(regfile[rs2.uint])), imm_B.int, returnPC))
    elif opcode == Opcode.LOAD:
        ALUOut = int(arithmetic(Funct3.ADD, regfile[rs1.uint], imm_I.int)) 
    elif opcode == Opcode.STORE:
        ALUOut = int(arithmetic(Funct3.ADD, regfile[rs1.uint], imm_S.int))
    elif opcode == Opcode.IMM:
        funct3 = Funct3(funct3)
        if funct3 in {Funct3.SLLI, Funct3.SRLI, Funct3.SRAI}:
            funct7 = Funct7(funct7)
            ALUOut = int(arithmetic(funct3=funct3, funct7=funct7, src1=regfile[rs1.uint], src2=rs2.uint))
        else:
            funct7 = None
            ALUOut = int(arithmetic(funct3=funct3, funct7=funct7, src1=regfile[rs1.uint], src2=imm_I.int))
    elif opcode == Opcode.OP:
        funct3 = Funct3(funct3)
        funct7 = Funct7(funct7)
        ALUOut = int(arithmetic(funct3=funct3, funct7=funct7, src1=regfile[rs1.uint], src2=regfile[rs2.uint]))
    elif opcode == Opcode.MISCMEM:
        return None 
    elif opcode == Opcode.SYSTEM:
        funct3 = Funct3(funct3)
        if funct3 != Funct3.ECALL:
            return None 
        if funct3 == Funct3.ECALL:
        # hack for test exit
            return False
    return int(ALUOut)


def memory_access(opcode, funct3, ALUOut, src2):
    if opcode == Opcode.STORE:
        MEMregister = None
        if funct3 == Funct3.SB:
            loader((src2 & 0xff).to_bytes(1,byteorder='little'), ALUOut)
        elif funct3 == Funct3.SH: 
            loader((src2 & 0xffff).to_bytes(2,byteorder='little'), ALUOut)
        elif funct3 == Funct3.SW:
            loader((src2 & 0xffffffff).to_bytes(4,byteorder='little') , ALUOut)
        return MEMregister
    elif opcode == Opcode.LOAD: 
        data = fetch(ALUOut)
        if funct3 == Funct3.LB:
            MEMregister = sign_extend(data[0], data & '0x000000FF') 
        elif funct3 == Funct3.LBU:
            MEMregister = sign_extend(0, data & '0x000000FF') 
        elif funct3 == Funct3.LHU:
            MEMregister = sign_extend(0, data & '0x0000FFFF') 
        elif funct3 == Funct3.LH:
            MEMregister = sign_extend(data[0], data & '0x0000FFFF') 
        elif funct3 == Funct3.LW:
            MEMregister = sign_extend(data[0], data & '0xFFFFFFFF')
        return MEMregister.int


def write_back(rd, write_data):
    regfile[rd.uint] = write_data 


def cycle() -> bool:
    # Fetch
    instruction = fetch(regfile[PC])
    #print(instruction)
    
    # Decode
    opcode, rd, src1, src2, funct3, funct7, I, S, B, U, J = decode(instruction) 
    #print("Opcode:", opcode, "rd:", rd, "rs1", src1, "rs2:", src2, "funct3:", funct3, "funct7:", funct7, "imm_I:", I, "imm_S:", S, "imm_B:", B, "imm_U:", U, "imm_J:", J)
    returnPC = regfile[PC] # set up return address for PC
    PCNext = regfile[PC] + 4
    mem_op = opcode in {Opcode.LOAD, Opcode.STORE} # check if need to store in memory
    write_op = opcode in {Opcode.LOAD, Opcode.OP, Opcode.IMM, Opcode.LUI, Opcode.JAL, Opcode.JALR, Opcode.AUIPC}  # check if need to write to memory
 
    # Execute
    ALUOut = execute(opcode, src1, src2, funct3, funct7, I, S, B, U, J, returnPC)
    #if ALUOut != None: print('ALUOut', hex(ALUOut)) 
    
    # Memory access
    if mem_op:
      MEMregister = memory_access(opcode, Funct3(funct3), ALUOut, regfile[src2.uint]) 
    
    # Write back
    if write_op:
        if opcode == Opcode.LOAD:
            write_back(rd, MEMregister)
        elif opcode in {Opcode.JAL, Opcode.JALR}:
            write_back(rd, regfile[PC]+4)
        else:
            write_back(rd, ALUOut)
    #hart_dump()

    # Calculate PC
    if opcode in {Opcode.BRANCH, Opcode.JAL, Opcode.JALR}:
       regfile[PC] = ALUOut   
    else:
        regfile[PC] = PCNext 
    
    # test if program is done 
    if ALUOut == False and type(ALUOut) == bool:
        return False
    else:
        return True


if __name__ == "__main__":
    for test_file in sorted(glob.glob("riscv-tests/isa/rv32ui-p-*")):
        if test_file.endswith('.dump') or test_file.endswith('fence_i'):
            continue
        with open(test_file, 'rb') as f:
            elf_file = ELFFile(f)
            print(test_file, ":")
            reset() # reset memory and PC for next program 
            regfile[PC] = 0x80000000  
            # ELF loader loads program into memory at location 0x80000000 
            for segment in elf_file.iter_segments(): 
                loader(segment.data(), segment.header.p_paddr) 
            instruction_count = 0 
            regfile[PC] = 0x80000174 # skip straight to instruction tests, ignoring csr instructions 
            while cycle():
                instruction_count += 1
             
            instruction = fetch(regfile[PC])
            print(hex(regfile[PC]))
            print(instruction.hex)
            if instruction.hex == 'c0001073':
                print("PASS") 
            else:
                print("FAIL")
            print("ran %d instructions\n" % instruction_count) 
            #exit(0)

File: rvis.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <elf.h>

//custom headers
#include "elfreader.h" // contains get_segment_data() and Elfinfo data type
#include "rv32.h" // bitfields data type 
#include "bitmanip.h" // contains sign_extend() and get_bit()

#define PC 32 // program counter is index 32 in regfile
#define MEM_SZ  0x1000000 // 16 MB of memory

// GLOBAL VARIABLES
byte memory[MEM_SZ]; // starts at 0x80000000
word regfile[33]; // 32, 32 bit word general purpose registers for RV32 and 1 32 bit register for the program counter

void reset() {
   memset(memory, 0, sizeof(memory)); 
   memset(regfile, 0, sizeof(regfile));
}

void hart_dump() {
    char reg_names[][33] = {
        "x0", "ra", "sp", "gp", "tp", "t0", "t1", "t2", 
        "s0", "s1", "a0", "a1", "a2", "a3", "a4", "a5",
        "a6", "a7", "s2", "s3", "s4", "s5", "s6", "s7",
        "s8", "s9", "s10", "s11", "t3", "t4", "t5", "t6",
        "PC" };
    for (int i=0; i<33; ++i) {
        if (i%6 == 0) 
            puts("\n");
        printf("%s:0x%08x\t", reg_names[i], regfile[i]);
        if (i == 32)
            puts("\n\n");
    }
}

void loader(ELFinfo elf, byte* segment_data, unsigned short segment_num) {
    Elf32_Addr address = elf.pheader[segment_num].p_paddr; 
    address -= elf.header.e_entry; // calculate offset
    if (address < 0 || address > MEM_SZ) {
         puts("LOAD FAIL"); 
         exit(EXIT_FAILURE);
    }
    printf("Address: %x\n", address); 
    for (size_t i=0; i < elf.pheader[segment_num].p_filesz; ++address, ++i) {
        memory[address] = segment_data[i];
   }
   puts("\n");
}

word fetch(word address, Elf32_Addr entry) {
    //word address = regfile[PC]; 
    address -= entry; // calculate offset to index memory array
    word instruction = 0; // instruction of length 32 bits
    word instruction_bytes[4]; // instructions come in 4 bytes for 32 bit arch
    if (address < 0 || address > MEM_SZ) {
        puts("FETCH FAILED"); 
        exit(EXIT_FAILURE);
    }
    for (int i = 0; i < 4; ++i) {
        instruction_bytes[i] = memory[address + i];
        instruction_bytes[i] = instruction_bytes[i] << (i*8); // shift i bytes 
        instruction += instruction_bytes[i]; 
    }
    //printf("Instruction: %x\n", instruction);
    return instruction;
}

bitfields decode(word instruction) {
// decode instruction following RISC-V 32 bit instruction format
    bitfields encoding = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    encoding.opcode = instruction & OPCODE;
    encoding.rd = (instruction & RD) >> 7;
    encoding.funct3 = (instruction & FUNCT3) >> 12;
    encoding.rs1 = (instruction & RS1) >> 15;
    encoding.rs2 = (instruction & RS2) >> 20;
    encoding.funct7 = (instruction & FUNCT7) >> 25;
    encoding.Itype = sign_extend((instruction & ITYPE) >> 20, 12);
    encoding.Stype = sign_extend(((instruction & STYPE_U) >> 20) |  ((instruction & STYPE_L) >> 7), 12);
    encoding.Btype = sign_extend((get_bits(1,32,instruction)<<12 | get_bits(1,8,instruction)<<11 | get_bits(6,26,instruction)<<5 | get_bits(4,9,instruction)<<1), 13);
    encoding.Utype = instruction & UTYPE;
    encoding.Jtype = sign_extend((get_bits(1,32,instruction)<<20 | get_bits(8,13,instruction)<<12 | get_bits(1,21,instruction)<<11 | get_bits(10,22,instruction)<<1), 21);
    return encoding;
}

word arithmetic(word funct3, bitfields encoding, word src2) { // src2 is either a 2nd register or an immediate 
    switch (funct3) {
        case F3_ADD: 
            if (encoding.funct7 == F7_SUB)
                return regfile[encoding.rs1] - src2;
            else return regfile[encoding.rs1] + src2;
        case F3_AND:
            return regfile[encoding.rs1] & src2;
        case F3_OR:
            return regfile[encoding.rs1] | src2;
        case F3_XOR:
            return regfile[encoding.rs1] ^ src2;
        case F3_SRL:
            if (encoding.funct7 == F7_SRL)
                return regfile[encoding.rs1] >> (src2 % 32);
            else
                return (int32_t)regfile[encoding.rs1] >> (src2 % 32); // sra and srai
        case F3_SLL:
            return regfile[encoding.rs1] << (src2 % 32);
        case F3_SLT:
            if ((int32_t)regfile[encoding.rs1] < (int32_t)src2)
                return 1;
            else return 0;
        case F3_SLTU:
            if (regfile[encoding.rs1] < src2)
                return 1;
            else return 0; 
        default:
            return 0;
    }
}

word logic(bitfields encoding) {
    switch (encoding.funct3) {
        case F3_BEQ:
            if ((int32_t)regfile[encoding.rs1] == (int32_t)regfile[encoding.rs2])
                return regfile[PC] + encoding.Btype;
            else return regfile[PC] + 4;
        case F3_BNE:
            if ((int32_t)regfile[encoding.rs1] != (int32_t)regfile[encoding.rs2])
                return regfile[PC] + encoding.Btype;
            else return regfile[PC] + 4;
        case F3_BLT:
            if ((int32_t)regfile[encoding.rs1] < (int32_t)regfile[encoding.rs2])
                return regfile[PC] + encoding.Btype;
            else return regfile[PC] + 4;
        case F3_BGE:
            if ((int32_t)regfile[encoding.rs1] >= (int32_t)regfile[encoding.rs2])
                return regfile[PC] + encoding.Btype;
            else return regfile[PC] + 4;
        case F3_BLTU:
            if (regfile[encoding.rs1] < regfile[encoding.rs2])
                return regfile[PC] + encoding.Btype;
            else return regfile[PC] + 4;
        case F3_BGEU:
            if (regfile[encoding.rs1] >= regfile[encoding.rs2])
                return regfile[PC] + encoding.Btype;
            else return regfile[PC] + 4;
        default:
            return 0;
    }
}

word execute(bitfields encoding) {
    switch (encoding.opcode) {
        case LUI:
            return (int32_t)encoding.Utype; 
             
        case AUIPC:
            //return regfile[PC] + (int32_t)encoding.Utype; 
        
        case JAL:
            return regfile[PC] + (int32_t)encoding.Jtype; 
        
        case JALR:
            return arithmetic(encoding.funct3, encoding, encoding.Itype);
        
        case BRANCH:
            return logic(encoding); 
        
        case LOAD:
            return arithmetic(F3_ADD, encoding, encoding.Itype);

        case STORE:
            return arithmetic(F3_ADD, encoding, encoding.Stype); 
        
        case IMM:
            if (encoding.funct3 == F3_SLLI || encoding.funct3 == F3_SRLI || encoding.funct3 == F3_SRAI)
                return arithmetic(encoding.funct3, encoding, encoding.rs2); 
            else return arithmetic(encoding.funct3, encoding, encoding.Itype);
        
        case OP:
            return arithmetic(encoding.funct3, encoding, regfile[encoding.rs2]);  
        
        case MISCMEM:
            return 0; 

        case SYSTEM:
            return 0;  
        
        default:
            return 0;
    } 
}

void load_byte(ELFinfo elf, word address, byte ALUout) {
    address -= elf.header.e_entry; // calculate offset
    if (address < 0 || address > MEM_SZ) {
        puts("LOAD_BYTE FAILED"); 
        exit(EXIT_FAILURE);
    }
    memory[address] = ALUout; 
}

word memory_access(ELFinfo elf, bitfields encoding, word ALUout) {
    word MEMregister = 0; 
    if (encoding.opcode == STORE) {
        if (encoding.funct3 == F3_SB) 
           load_byte(elf, ALUout, (regfile[encoding.rs2] & 0xFF)); 
         
        else if (encoding.funct3 == F3_SH) 
            for (unsigned char i=0; i<4; ++i)
                load_byte(elf, ALUout + i, ((regfile[encoding.rs2] & (0xFF << i*8)) >> i*8));
        
        else if (encoding.funct3 == F3_SW) 
            for (unsigned char i=0; i<4; ++i)
                load_byte(elf, ALUout + i, ((regfile[encoding.rs2] & (0xFF << i*8)) >> i*8));
    }
    else {
        word data = fetch(ALUout, elf.header.e_entry);
        printf("DATA: %x\n", data);
        if (encoding.funct3 == F3_LB)
            MEMregister = sign_extend((int32_t)data & 0xFF, 8);  
        
        else if (encoding.funct3 == F3_LBU)
            MEMregister = (byte)data;
        
        else if (encoding.funct3 == F3_LHU)
            MEMregister = data & 0xFFFF;
        
        else if (encoding.funct3 == F3_LH)
            MEMregister = sign_extend((int32_t)data & 0xFFFF, 16);
        
        else if (encoding.funct3 == F3_LW) 
            MEMregister = (int32_t)data; 
    }
    return MEMregister;
    
}

void write_back(bitfields encoding, word ALUout) {
    regfile[encoding.rd] = ALUout;
    regfile[0] = 0; // x0 is always 0 
}

bool cycle(ELFinfo elf) {
    // fetch  
    word instr = fetch(regfile[PC], elf.header.e_entry);
    printf("Instr: %x\n", instr);    
    
    // decode
    bitfields encoding = decode(instr);
    printf("Opcode: 0x%02x rd: 0x%02x rs1: 0x%02x rs2:  0x%02x funct3: 0x%01x funct7: 0x%02x\n", 
        encoding.opcode, encoding.rd, encoding.rs1, encoding.rs2, encoding.funct3,  encoding.funct7);
    printf("I: 0x%08x S: 0x%08x B: 0x%08x U: 0x%08x J: 0x%08x\n",
        encoding.Itype, encoding.Stype, encoding.Btype, encoding.Utype, encoding.Jtype);
     
    // execute
    word ALUout = execute(encoding);
    printf("ALUout: %x\n", ALUout);   
    
    // memory access
    word MEMregister = 0;
    if (encoding.opcode == LOAD || encoding.opcode == STORE){ 
        MEMregister = memory_access(elf, encoding, ALUout);
        printf("MEMregister: %x\n", MEMregister);
    }

    // write back
    if (encoding.opcode == LOAD || encoding.opcode == OP || encoding.opcode == IMM || 
        encoding.opcode == LUI || encoding.opcode == JAL || encoding.opcode == JALR || encoding.opcode == AUIPC) {
        
        if (encoding.opcode == LOAD) 
            write_back(encoding, MEMregister);
        
        else if (encoding.opcode == JAL || encoding.opcode == JALR) 
            write_back(encoding, regfile[PC] + 4);
        
        else {
            write_back(encoding, ALUout);
        }
    }
    
    // show register content
    hart_dump();
    
    // PC calculation
    if (encoding.opcode == BRANCH || encoding.opcode == JAL || encoding.opcode == JALR)
        regfile[PC] = ALUout;
    else regfile[PC] += 4;
    
    if (encoding.opcode == SYSTEM && encoding.funct3 == F3_ECALL) 
        return false;
    else return true;
}

int main(){
    char elf_file[] = "riscv-tests/isa/rv32ui-p-sh";
    ELFinfo elf = read_elf(elf_file);
    byte* segments_data[elf.header.e_phnum];
    for (size_t i = 0; i<elf.header.e_phnum; ++i) {
        segments_data[i] = get_segment_data(elf, i); 
        loader(elf, segments_data[i], i);
    }
    /* 
    for (int i=0; i<elf.pheader[0].p_filesz; ++i) 
        printf("%0hhx", memory[i]); 
    puts("\n"); */
//    for (int i=0; i<elf.pheader[1].p_filesz; ++i) 
//        printf("%0hhx", memory[elf.pheader[0].p_filesz + i]); 
//    puts("\n");
   

//    for (int i=0; i<elf.pheader[1].p_filesz; ++i)
//        printf("%0hhx", segments_data[1][i]);
//    puts("\n");
    regfile[PC] = 0x80000174;
    unsigned int instr_count = 0; 
    while(cycle(elf)){
        instr_count += 1;
    }
    //word instr = fetch(regfile[PC], elf.header.e_entry);
    //printf("PC: %x\n", regfile[PC]); 
    //printf("INSTR: %x\n", instr);
    printf("Ran: %d instructions\n", instr_count);
    //printf("MEMORY: %x\n", memory[0x1000]); 
    //for (int i=0; i<0x2050; i++)
    //    printf("ADDRESS: %x    Value: %x\n", i+0x80000000, memory[i]);

    free(elf.pheader);
    free(elf.buffer);
    for (int i=0; i<elf.num_pheaders; ++i)
        free(segments_data[i]);
    return EXIT_SUCCESS;
}
